batched_blas/batched_blas.md

```diff
@@ -325,6 +325,110 @@ have a natural representation without them.
+
+template<class T>
+concept @_out-vector_@ = // exposition only
+  @_is-mdspan_@<T>::value &&
+  T::rank() == 1 &&
+  is_assignable_v<typename T::reference, typename T::element_type> &&
+  T::is_always_unique();
+```
+
+We propose adding new overloads of the function with new exposition only concepts, accounting for broadcasting of input arguments:
+
+```c++
+template<class T>
+concept @_batched-in-vector_@ = // exposition only
+  @_is-mdspan_@<T>::value &&
+  (T::rank() == 1 || T::rank() == 2);
+
+template<class T>
+concept @_batched-out-vector_@ = // exposition only
+  @_is-mdspan_@<T>::value &&
+  T::rank() == 2 &&
+  is_assignable_v<typename T::reference, typename T::element_type> &&
+  T::is_always_unique();
+```
+
+The batched _`batched-out-***`_ and _`batched-inout-***`_ concepts strictly increase the rank requirement by one, _`in-***`_ concepts allow the original rank and one higher.
+
+In addition to these new exposition only concepts we need some helper functions:
+
+```c++
+template<@_batched-in-vector_@ InVec>
+requires(InVec::rank()==1)
+auto get_batch(InVec v, int /*batch*/) { return v; }
+
+template<@_batched-in-vector_@ InVec>
+requires(InVec::rank()==2)
+auto get_batch(InVec v, int batch) { return submdspan(v, batch, full_extent); }
+
+template<@_batched-out-vector_@ OutVec>
+auto get_batch(OutVec v, int batch) { return submdspan(v, batch, full_extent); }
+```
+
+With those functions we can now define the batched overload for `matrix_vector_product`:
+
+```c++
+template<class ExecutionPolicy,
+         @_batched-in-matrix_@ InMat,
+         @_batched-in-vector_@ InVec,
+         @_batched-out-vector_@ OutVec>
+void matrix_vector_product(ExecutionPolicy&& exec,
+                           InMat A,
+                           InVec x,
+                           OutVec y);
+```
+
+*Preconditions:* `A.extent(0) == x.extent(0) && A.extent(0) == y.extent(0)` is `true`.
+
+*Effects:* Equivalent to calling `matrix_vector_product(get_batch(A, i), get_batch(x,  i), get_batch(y, i))` for each `i` in the range of $[$ 0,`A.extent(0)`$)$. 
+
+This leaves the question of having calls with separate scaling factor per subobject.
+To support that we add an additional overload taking a rank-1 `mdspan` with the scaling factors:
+
+```c++
+template<class ExecutionPolicy,
+         @_in-vector_@ Alphas,
+         @_batched-in-matrix_@ InMat,
+         @_batched-in-vector_@ InVec,
+         @_batched-out-vector_@ OutVec>
+void matrix_vector_product(ExecutionPolicy&& exec,
+                           Alphas alphas,
+                           InMat A,
+                           InVec x,
+                           OutVec y);
+```
+
+*Preconditions:* `A.extent(0) == x.extent(0) && A.extent(0) == y.extent(0) && A.extent(0) == alphas.extent(0)` is `true`.
+
+*Effects:* Equivalent to calling `matrix_vector_product(scaled(alphas[i],get_batch(A, i)), get_batch(x,  i), get_batch(y, i))` for each `i` in the range of $[$ 0,`A.extent(0)`$)$. 
```

Comment:
I don't think this works.

Reply:
as in you don't think this is the right way to word it?

Reply:
It was related to my comments above about overload resolution and get_batch, but I see your updated version has get_batch_{matrix,vector} so it looks better.